### ISSUE TITLE
Migrations: CHECK enum columns

### ISSUE THREAD

**Issue by AndriySvyryd** [aspnet/EntityFrameworkCore#15411]

**Comment by SARAVANA1501** [comment]
hi @AndriySvyryd, Could you tell me which file i can add this CHECK?

**Comment by AndriySvyryd** [comment]
This is about adding a CHECK constraint to enum columns e.g.:
```C#
mycol VARCHAR(10) NOT NULL CHECK (mycol IN('Useful', 'Useless', 'Unknown'))
```

The easiest way of doing this is by adding a SqlServer convention (IModelFinalizedConvention) that calls `AddCheckConstraint` for all enum properties.

**Comment by SARAVANA1501** [comment]
Hi @AndriySvyryd 
Following test is my expectation, Is that correct?
` [ConditionalFact]
        public void Can_create_check_constraint_for_enum()
        {
            var modelBuilder = CreateConventionModelBuilder();
            var entityType = modelBuilder.Entity<Customer>().Metadata;

            modelBuilder
                .Entity<Customer>()
                .HasCheckConstraint("CK_Customer_EnumValue", "CHECK (mycol IN('Sun', 'Mon', 'Tue'))");

            var checkConstraint = entityType.FindCheckConstraint("CK_Customer_EnumValue");

            Assert.NotNull(checkConstraint);
            Assert.Equal(entityType, checkConstraint.EntityType);
            Assert.Equal("CK_Customer_EnumValue", checkConstraint.Name);
            Assert.Equal("CHECK (EnumValue IN('Sun', 'Mon', 'Tue'))", checkConstraint.Sql);
        }`

**Comment by AndriySvyryd** [comment]
@SARAVANA1501 Not quite, this should happen without needing to call `WithEnumConstraint()`

**Comment by SARAVANA1501** [comment]
@AndriySvyryd , 

[SqlServerEnumConstraintConvention](https://github.com/SARAVANA1501/EntityFrameworkCore/blob/15411-check-enum/src/EFCore.SqlServer/Metadata/Conventions/SqlServerEnumConvention.cs)

Please review the convention, and guide me to register it with other conventions,

Thanks.

**Comment by AndriySvyryd** [comment]
You are heading in the right direction.

- Call `entityType.GetDeclaredProperties()` instead of `entityType.GetProperties()` to avoid going over the same properties.
- Don't call `property.SetIsNullable(false)`
- The convention can be added in SqlServerConventionSetBuilder:
```C#
    ConventionSet.AddBefore(
        conventionSet.ModelFinalizedConventions,
        new SqlServerEnumConvention(),
        typeof(ValidatingConvention));
```
- Add tests to EFCore.SqlServer.Tests

I'll give further feedback on the PR

**Comment by AndriySvyryd** [comment]
Fixed in e17584c457a49c4dcd90fd5cf47cd3a61ae4db53
